instance/models/mixins/rabbitmq.py

```diff
@@ -0,0 +1,115 @@
+#
+"""
+Instance app model mixins - RabbitMQ
+"""
+
+# Imports #####################################################################
+
+import pathlib
+
+from django.conf import settings
+from django.db import models
+
+from instance.ansible import capture_playbook_output
+from instance.models.rabbitmq import RabbitMQUser
+
+
+# Classes #####################################################################
+
+class ReconfigurationFailed(Exception):
+    """Exception indicating that reconfiguring RabbitMQ failed."""
+
+
+class RabbitMQInstanceMixin(models.Model):
+    """
+    An instance that uses a RabbitMQ vhost with a set of users.
+    """
+    rabbitmq_vhost = models.CharField(max_length=16, blank=True)
+    rabbitmq_provider_user = models.ForeignKey(
+        RabbitMQUser,
+        related_name='provider_instance',
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True
+    )
+    rabbitmq_consumer_user = models.ForeignKey(
+        RabbitMQUser,
+        related_name='consumer_instance',
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True
+    )
+    rabbitmq_provisioned = models.BooleanField(default=False)
+
+    class Meta:
+        abstract = True
+
+    def _get_rabbitmq_vars(self, remove):
+        """
+        Generate the vars string for the RabbitMQ playbook.
+        """
+        users = ''.join([
+            '- username: {username}\n  password: {password}\n'.format(username=user.username, password=user.password)
+            for user in [self.rabbitmq_provider_user, self.rabbitmq_consumer_user]
+        ])
+        ansible_vars = 'STATE: {state}\nRABBITMQ_VHOST: {vhost}\nRABBITMQ_USERS:\n{users}'.format(
+            state='absent' if remove else 'present',
+            vhost=self.rabbitmq_vhost,
+            users=users
+        )
+
+        return ansible_vars
+
+    def _run_rabbitmq_playbook(self, remove):
+        """
+        Run the RabbitMQ playbook to reconfigure add/remove a vhost and users.
+        """
+        playbook_path = pathlib.Path(settings.SITE_ROOT) / "playbooks/rabbitmq_conf/rabbitmq_conf.yml"
+        returncode = capture_playbook_output(
+            requirements_path=str(playbook_path.parent / "requirements.txt"),
+            inventory_str=settings.INSTANCE_RABBITMQ_HOST,
+            vars_str=self._get_rabbitmq_vars(remove),
+            playbook_path=str(playbook_path),
+            username=settings.INSTANCE_RABBITMQ_SSH_USERNAME,
+            logger_=self.logger,
+        )
+        if returncode != 0:
+            self.logger.error("Playbook to reconfigure RabbitMQ server %s failed.", self)
+            raise ReconfigurationFailed
+
+        return returncode
+
+    def provision_rabbitmq(self):
+        """
+        Create RabbitMQ vhost and users
+        """
+        if settings.RABBITMQ_ENABLE and not self.rabbitmq_provisioned:
+            self._run_rabbitmq_playbook(False)
+        self.rabbitmq_provisioned = True
```

Comment:
If `RABBITMQ_ENABLE` is False, we mark rabbitmq as provisioned even though we don't use it and haven't actually provisioned it?